subt/main.py

```diff
@@ -469,6 +469,32 @@ def on_pose2d(self, timestamp, data):
         self.xyz = x, y, z
         self.trace.update_trace(self.xyz)
 
+    def on_pose3d(self, timestamp, data):
+        xyz, rot = data
+        ypr = quaternion.euler_zyx(rot)
+        pose = (xyz[0], xyz[1], ypr[0])
+        if self.last_position is not None:
+            self.is_moving = (self.last_position != pose)
```

Comment:
One issue here is that last_position is initialized to (0, 0, 0) (i.e. is in robot-centric coordinate frame), while the 3d pose incorporates initial position of the robot relative to the global coordinate frame, i.e. it is world-centric.

The options we have are:
a) Accept that is_moving will be True in the beginning even when it shouldn't. (Current case.)
b) Make `pose` relative to robot's starting position (i.e. subtract `self.offset`).
c) Initialize `self.last_position` to None and make sure that everywhere we touch it, we check first if it is already not null. (Or somewhere in the beginning wait until it gets a value.)

What do you prefer?

Reply:
You are right. I think that the `self.last_position` is actually potentially bigger problem than `self.is_moving` used only to verify the robot is stopped after motion (it would add extra cycle). For example `go_straight` 1 meter with later update that we are at (5m, 8m) would immediately stop ... so I am inclined to use c) option. I would also double check that localization publishes coordinates only after the global frame is known ... with combination of real robots ... I can feel the tar pit (Brooks) :). It also means that you cannot move 1m until you know your global coordinate, which is kind of awkward.

Reply:
If you memorize the trajectory, in the global coordinate system, you shouldn't move before knowing that coordinate system.